itest/rust/src/builtin_tests/convert_test.rs

```diff
@@ -0,0 +1,241 @@
+    const TOO_MANY_KEYS: &'static str = "missing `a` key";
+}
+
+impl GodotConvert for Foo {
+    type Via = Dictionary;
+}
+
+impl ToGodot for Foo {
+    fn to_godot(&self) -> Self::Via {
+        dict! {
+            "a": self.a,
+            "b": self.b,
+        }
+    }
+}
+
+impl FromGodot for Foo {
+    fn try_from_godot(via: Self::Via) -> Result<Self, ConvertError> {
+        let a = match via.get("a") {
+            Some(a) => a,
+            None => {
+                return Err(ConvertError::custom()
+                    .with_cause(Self::MISSING_KEY_A)
+                    .with_value(via))
+            }
+        };
+
+        let b = match via.get("b") {
+            Some(b) => b,
+            None => {
+                return Err(ConvertError::custom()
+                    .with_cause(Self::MISSING_KEY_B)
+                    .with_value(via))
+            }
+        };
+
+        if via.len() > 2 {
+            return Err(ConvertError::custom()
+                .with_cause(Self::TOO_MANY_KEYS)
+                .with_value(via));
+        }
+
+        Ok(Self {
+            a: a.try_to()?,
+            b: b.try_to()?,
+        })
+    }
+}
+
+#[itest]
+fn custom_convert_roundtrip() {
+    let foo = Foo { a: 10, b: 12.34 };
+
+    let as_dict = foo.to_godot();
+    assert_eq!(as_dict.get("a"), Some(foo.a.to_variant()));
+    assert_eq!(as_dict.get("b"), Some(foo.b.to_variant()));
+
+    let foo2 = as_dict.to_variant().to::<Foo>();
+    assert_eq!(foo, foo2, "from_variant");
+
+    let foo3 = Foo::from_godot(as_dict);
+    assert_eq!(foo, foo3, "from_godot");
+}
+
+// Ensure all failure states for the `FromGodot` conversion of `Foo` are propagated through the `try_to`
+// method of `Variant` as they should be.
+#[itest]
+fn custom_convert_error_from_variant() {
+    let missing_a = dict! {
+        "b": -0.001
+    };
+    let err = missing_a
+        .to_variant()
+        .try_to::<Foo>()
+        .expect_err("should be missing key `a`");
+    assert_eq!(err.cause().unwrap().to_string(), Foo::MISSING_KEY_A);
+
+    let missing_b = dict! {
+        "a": 58,
+    };
+    let err = missing_b
+        .to_variant()
+        .try_to::<Foo>()
+        .expect_err("should be missing key `b`");
+    assert_eq!(err.cause().unwrap().to_string(), Foo::MISSING_KEY_B);
+
+    let too_many_keys = dict! {
+        "a": 12,
+        "b": 777.777,
+        "c": "bar"
+    };
+    too_many_keys
+        .to_variant()
+        .try_to::<Foo>()
+        .expect_err("should have too many keys");
+    assert_eq!(err.cause().unwrap().to_string(), Foo::TOO_MANY_KEYS);
+
+    let wrong_type_a = dict! {
+        "a": "hello",
+        "b": 28.41,
+    };
+    let err = wrong_type_a
+        .to_variant()
+        .try_to::<Foo>()
+        .expect_err("should have wrongly typed key `a`");
+    assert!(err.cause().is_none());
+    assert_eq!(
+        format!("{:?}", err.value().unwrap()),
+        format!("{:?}", "hello".to_variant())
```

Comment:
It's not obvious why you compare the `Debug` reprs here -- is it because one can't fetch/downcast the original type anymore? Maybe add a comment 🤔 

(also in other such cases)

Reply:
oh yeah i did add a comment earlier in another case but i can add it here too